crates/propolis-api-types/src/instance_spec/mod.rs

```diff
@@ -187,29 +186,20 @@ pub mod v0;
 // names if they have no UUIDs available or the most obvious UUID is in use
 // elsewhere. The key type's From impls will try to parse strings into UUIDs
 // before storing keys as strings.
-//
-// This type derives `SerializeDisplay` and `DeserializeFromStr` so that it can
-// be used as a map key when serializing to JSON, which requires strings (and
-// not objects) as keys.
-//
-// WARNING: Progenitor-generated versions of this type will use serde's default
-// derived Serialize and Deserialize implementations, not the serde_with
-// implementations. This can cause clients to serialize spec keys in a form that
-// the server can't deserialize. To get around this, Progenitor clients that
-// use spec keys should use a `replace` directive to replace the generated
-// type with this type.
 #[derive(
-    Clone,
-    Debug,
-    SerializeDisplay,
-    DeserializeFromStr,
-    Eq,
-    PartialEq,
-    Ord,
-    PartialOrd,
-    JsonSchema,
+    Clone, Debug, Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd,
 )]
-#[serde(rename_all = "snake_case")]
+// Direct serde to use an untagged enum representation for this type. Since both
+// Uuid and String serialize to strings, this allows other types that contain a
+// Map<K = SpecKey> to derive Serialize and successfully serialize to JSON.
+// (This doesn't work with a tagged representation because JSON doesn't allow
+// maps to be used as map keys.)
+//
+// Note that this makes the order of variants matter: serde will pick the first
+// variant into which it can successfully deserialize an untagged enum value,
+// and the point is to use the UUID representation for any value that can be
+// interpreted as a UUID.
```

Comment:
Ugh, that's ugly, but...not the end of the world, I guess... Perhaps we should add a test that would fail if someone changed the order of the variants, just in case?

Reply:
The existing unit tests in this module cover this (past-me looking out for present-me for once!). Specifically they check that (a) serializing a `SpecKey::Uuid` and deserializing it produces another `SpecKey::Uuid`, and (b) serializing a `SpecKey::Name(Uuid::new_v4().to_string())` and deserializing that also gives you a `SpecKey::Uuid`. Both tests break if the `Uuid` variant isn't first in this enum.